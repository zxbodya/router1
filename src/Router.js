import compileRoutes from './compileRoutes';
import {Subject} from 'rx';

class Router {
  constructor({history, routes, render}) {
    this.history = history;
    this.routes = [];
    this.render = render;

    this.routesByName = {};
    this.addRoutes(routes);

    this.activeRoute = [null, {}];

    this.hashChange = new Subject();
    this.currentLocation = {};
  }

  addRoutes(routeDefs) {
    compileRoutes(routeDefs)
      .forEach(route=> {
        this.routes.push(route);
        this.routesByName[route.name] = route;
      });
  }

  matchRoute(location) {
    const {pathname} = location;
    const matched = [];

    for (let i = 0, l = this.routes.length; i < l; i++) {
      const route = this.routes[i];
      const params = route.matchPath(pathname);
      if (params) {
        matched.push([route, params]);
      }
    }

    let res;
    if (matched.length === 0) {
      // todo: route not found clause
      console.log('route not found');
    } else {
      res = matched[0];

      // todo: conflict clause
      if (matched.length > 1) {
        console.log('matched few routes');
      }
      // 1.  warning in dev mode
      // 2.  optional data resolving step
      // 2.1 select from conflicting
      // 2.2 resource not found clause
    }

    if (res) {
      const route = res[0];
      const params = res[1];

      return {route: route.name, handler: route.handler, params, location};
    }
    return {route: null, handler: null, params: {}, location};
  }

  renderResult() {
    return this.history
      .location
      .filter(location=> {
        let needUpdate = true;
        // todo: search
        if (this.currentLocation.pathname === location.pathname) {
          this.hashChange.onNext(location);
          needUpdate = false;
        }
        this.currentLocation = location;
        return needUpdate;
      })
      .map(this.matchRoute.bind(this))
      .do(({route, params})=> {
        this.activeRoute = [route, params];
      })
      .flatMapLatest(this.render)
      .share();
  }

  isActive(route, params) {
    if (this.activeRoute[0] && route === this.activeRoute[0]) {
      let active = true;

      let paramName;
      for (paramName in params) {
        if (params.hasOwnProperty(paramName)) {
          active = active && params[paramName].toString() === this.activeRoute[1][paramName].toString();
        }
      }
      return active;
    }
    return false;
  }

  createUrl(name, params = {}, hash = '') {
    const route = this.routesByName[name];
    if (route) {
      const pathname = route.generatePath(Object.assign({}, this.activeRoute[1], params));
      return `${pathname}${hash ? `#${hash}` : ''}`;
    }
    console.error(`Route "${name}" not found`);
    return `#route-${name}-not-found`;
  }

  navigate(route, params = {}, hash = '') {
    const url = this.createUrl(route, params, hash);
    this.history.push(url);
  }

  navigateToUrl(url) {
    this.history.push(url);
  }

}

export default Router;
