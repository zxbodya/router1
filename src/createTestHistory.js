import { Subject } from 'rxjs/Subject';
import 'rxjs/add/operator/startWith';
import 'rxjs/add/operator/publishReplay';


import { locationFromUrl } from './utils/locationFromUrl';

export function createTestHistory(initialUrl, cb) {
  const location$ = new Subject();
  return {
    createUrl(pathname, search, hash) {
      return `${pathname}${search ? `?${search}` : ''}${hash ? `#${hash}` : ''}`;
    },
    parseUrl(url) {
      return locationFromUrl(url);
    },
    push(url, state = null, title = null) {
      if (cb) cb('push', { url, state, title });
    },
    replace(url, state = null, title = null) {
      if (cb) cb('replace', { url, state, title });
    },
    navigate(url, state) {
      location$.next(
        Object.assign(locationFromUrl(url, state), { source: 'pop' })
      );
    },
    location: location$
      .startWith(
        Object.assign(locationFromUrl(initialUrl), { source: 'init' })
      )
      .publishReplay(1)
      .refCount(),
  };
}
