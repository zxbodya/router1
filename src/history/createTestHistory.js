import { Subject } from 'rxjs/Subject';
import { startWith } from 'rxjs/operators/startWith';
import { publishReplay } from 'rxjs/operators/publishReplay';
import { refCount } from 'rxjs/operators/refCount';

import { locationFromUrl } from '../utils/locationFromUrl';

export function createTestHistory(initialUrl, cb) {
  const location$ = new Subject();
  return {
    createUrl(path, search, hash) {
      return `${path}${search ? `?${search}` : ''}${hash ? `#${hash}` : ''}`;
    },
    parseUrl(url) {
      return locationFromUrl(url);
    },
    push(url, state = null, title = null) {
      if (cb) cb('push', { url, state, title });
    },
    replace(url, state = null, title = null) {
      if (cb) cb('replace', { url, state, title });
    },
    navigate(url, state) {
      location$.next(
        Object.assign(locationFromUrl(url, state), { source: 'pop' })
      );
    },
    location: location$.pipe(
      startWith(Object.assign(locationFromUrl(initialUrl), { source: 'init' })),
      publishReplay(1),
      refCount()
    ),
  };
}
