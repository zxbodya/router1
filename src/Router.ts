import { EMPTY, merge, noop, Observable, of, Subject } from 'rxjs';

import { map, mergeMap, switchMap } from 'rxjs/operators';

import { Subscriber, Subscription } from 'rxjs';

import { History, Location, NavigateSource } from './history/history';
import { Route } from './routes/compileRoutes';
import { RouteCollection } from './routes/RouteCollection';

import {
  generate as generateQuery,
  parse as parseQuery,
} from './utils/queryString';

import { normalizeParams } from './routes/normalizeParams';
import { OnLocationChange, StateHandler } from './StateHandler';

export interface Transition<RouteState, RenderResult, RouteHandler> {
  location: Location;
  // eslint-disable-next-line no-use-before-define
  router: Router<RouteState, RenderResult, RouteHandler>;
  redirectCount: number;
  forward: (redirectUrl: string) => void;
}

export interface RouteParams {
  [key: string]: string | boolean;
}
export type RouteTransition<
  RouteState,
  RenderResult,
  RouteHandler
> = Transition<RouteState, RenderResult, RouteHandler> & {
  route: Route<RouteHandler>;
  params: RouteParams;
};

type StateLoader<RouteState, RenderResult, RouteHandler> = ((
  routeTransition: RouteTransition<RouteState, RenderResult, RouteHandler>
) => Observable<RouteState>);

export type AfterRender<RouteState, RenderResult, RouteHandler> = (
  stateHandler: StateHandler<RouteState, RenderResult, RouteHandler>,
  pageState: {
    state: RouteState;
    transition: Transition<RouteState, RenderResult, RouteHandler>;
    renderResult: RenderResult;
  }
) => void;

export interface RouterConfig<RouteState, RenderResult, RouteHandler> {
  history: History;
  routeCollection: RouteCollection<RouteHandler>;
  loadState: StateLoader<RouteState, RenderResult, RouteHandler>;
  onHashChange?: OnLocationChange;
  renderState: (
    state: RouteState,
    transition: RouteTransition<RouteState, RenderResult, RouteHandler>
  ) => Observable<RenderResult>;
  onLocationChange?: OnLocationChange;
}

export type RenderState<RouteState, RenderResult, RouteHandler> = (
  state: RouteState,
  routeTransition: RouteTransition<RouteState, RenderResult, RouteHandler>
) => Observable<RenderResult>;

export class Router<RouteState, RenderResult, RouteHandler> {
  private readonly history: History;

  private readonly routeCollection: RouteCollection<RouteHandler>;

  public activeRoute: [
    string,
    RouteParams,
    StateHandler<RouteState, RenderResult, RouteHandler> | null
  ] = ['', {}, null];
  public currentLocation: Location = {
    pathname: '',
    search: '',
    hash: '',
    state: null,
  };

  private resultsSubscription?: Subscription = undefined;

  private readonly renderResult$: Subject<RenderResult> = new Subject();

  private readonly loadState: StateLoader<
    RouteState,
    RenderResult,
    RouteHandler
  >;

  private readonly onHashChange: OnLocationChange;

  private readonly renderState: RenderState<
    RouteState,
    RenderResult,
    RouteHandler
  >;

  private readonly onLocationChange: OnLocationChange;

  private readonly navigate$: Subject<{
    url: string;
    state?: object;
    source: NavigateSource;
  }> = new Subject();

  public onBeforeUnload = (e?: BeforeUnloadEvent): string | void => {
    const returnValue = this.activeRoute[2]
      ? this.activeRoute[2].onBeforeUnload()
      : '';
    if (e && returnValue) {
      e.returnValue = returnValue;
    }
    return returnValue || undefined;
  };

  constructor(config: RouterConfig<RouteState, RenderResult, RouteHandler>) {
    this.history = config.history;
    this.routeCollection = config.routeCollection;
    this.loadState = config.loadState;
    this.onHashChange = config.onHashChange || noop;
    this.renderState = config.renderState;
    this.onLocationChange = config.onLocationChange || noop;
  }

  private createHandler(
    transition: Transition<RouteState, RenderResult, RouteHandler> & {
      route: Route<RouteHandler>;
      params: RouteParams;
    }
  ): Observable<StateHandler<RouteState, RenderResult, RouteHandler> | null> {
    const state$ = this.loadState(transition);
    return state$.pipe(
      map(
        state =>
          state
            ? new StateHandler(
                state,
                transition,
                this.onHashChange,
                this.renderState,
                this.onLocationChange
              )
            : null
      )
    );
  }

  private createNotFoundHandler(
    transition: Transition<RouteState, RenderResult, RouteHandler>
  ): Observable<StateHandler<RouteState, RenderResult, RouteHandler>> {
    const notFoundTransition = {
      // @ts-ignore
      route: {
        name: '',
        handlers: [],
      } as Route<RouteHandler>,
      params: {},
      ...transition,
      router: this,
    };
    const state$ = this.loadState(notFoundTransition);
    return state$.pipe(
      map(
        state =>
          new StateHandler(
            state,
            notFoundTransition,
            this.onHashChange,
            this.renderState,
            this.onLocationChange
          )
      )
    );
  }

  public start(): void {
    const transitionFromLocation = (
      toLocation: Location
    ): Observable<Transition<RouteState, RenderResult, RouteHandler>> =>
      Observable.create(
        (
          observer: Subscriber<
            Transition<RouteState, RenderResult, RouteHandler>
          >
        ) => {
          let redirectCount = 0;
          let forwardInt: (url: string) => void;
          // defer redirect to new state to prevent subscription new render() result before old in edge cases
          const forward = (redirectUrl: string) => {
            setTimeout(forwardInt, 0, redirectUrl);
          };
          forwardInt = (redirectUrl: string): void => {
            if (redirectCount > 20) {
              observer.error(Error('To many redirects!'));
            }

            this.history.replace(redirectUrl);

            const location: Location = {
              ...this.history.parseUrl(redirectUrl),
              source: 'replace',
              state: {},
            };

            if (
              this.currentLocation.pathname === location.pathname &&
              this.currentLocation.search === location.search
            ) {
              if (this.currentLocation.hash === location.hash) {
                observer.error(Error('Redirect to the same location!'));
              }
              if (this.activeRoute[2]) {
                this.activeRoute[2].onHashChange(location);
              }
              this.currentLocation = location;
            } else {
              this.currentLocation = location;

              redirectCount += 1;
              observer.next({
                location,
                router: this,
                redirectCount,
                forward,
              });
            }
          };
          observer.next({
            location: toLocation,
            router: this,
            redirectCount,
            forward,
          });
        }
      );

    const historyTransition$: Observable<Location> = this.history.location.pipe(
      mergeMap((location: Location) => {
        if (
          this.currentLocation.pathname === location.pathname &&
          this.currentLocation.search === location.search
        ) {
          if (this.activeRoute[2]) {
            this.activeRoute[2].onHashChange(location);
          }
          this.currentLocation = location;
          return EMPTY;
        }

        const beforeUnload = this.onBeforeUnload();
        // eslint-disable-next-line no-restricted-globals,no-alert
        const cancelTransition = beforeUnload && !confirm(beforeUnload);
        if (cancelTransition) {
          // case when user navigates back or forward, but transition was canceled
          this.history.push(
            this.history.createUrl(
              this.currentLocation.pathname,
              this.currentLocation.search,
              this.currentLocation.hash
            ),
            this.currentLocation.state
          );
          return EMPTY;
        }

        this.currentLocation = location;
        return [location];
      })
    );

    const navigateTransition$: Observable<Location> = this.navigate$.pipe(
      mergeMap(({ url, state, source }) => {
        const location = {
          ...this.history.parseUrl(url),
          source,
          state,
        };

        if (
          this.currentLocation.pathname === location.pathname &&
          this.currentLocation.search === location.search
        ) {
          if (this.activeRoute[2]) {
            this.activeRoute[2].onHashChange(location);
          }
          this.currentLocation = location;
          this.history.push(url, state);
          return EMPTY;
        }

        const beforeUnload = this.onBeforeUnload();
        // eslint-disable-next-line no-restricted-globals,no-alert
        const cancelTransition = beforeUnload && !confirm(beforeUnload);

        if (cancelTransition) {
          return EMPTY;
        }

        this.currentLocation = location;
        this.history.push(url, state);

        return [location];
      })
    );
    const matchRoutes = (
      transition: Transition<RouteState, RenderResult, RouteHandler>
    ): Transition<RouteState, RenderResult, RouteHandler> & {
      routes: Array<[Route<RouteHandler>, RouteParams]>;
    } => ({
      ...transition,
      routes: this.routeCollection.match(
        transition.location.pathname,
        parseQuery(transition.location.search)
      ),
    });

    const loadMatched = (
      transition: Transition<RouteState, RenderResult, RouteHandler> & {
        routes: Array<[Route<RouteHandler>, RouteParams]>;
      }
    ): Observable<
      [
        string,
        RouteParams,
        StateHandler<RouteState, RenderResult, RouteHandler>
      ]
    > => {
      const loadRoute = (
        routes: Array<[Route<RouteHandler>, RouteParams]>,
        index: number
      ): Observable<
        [
          string,
          RouteParams,
          StateHandler<RouteState, RenderResult, RouteHandler> | null
        ]
      > => {
        if (index >= routes.length) {
          // not found
          return of(['', {}, null]) as Observable<
            [
              string,
              RouteParams,
              StateHandler<RouteState, RenderResult, RouteHandler>
            ]
          >;
        }

        const route = routes[index];
        const handler = this.createHandler({
          route: route[0],
          params: route[1],
          ...transition,
        } as Transition<RouteState, RenderResult, RouteHandler> & { route: any; params: any });
        return handler.pipe(
          switchMap(
            loadResult =>
              loadResult
                ? (of([route[0].name, route[1], loadResult]) as Observable<
                    [
                      string,
                      RouteParams,
                      StateHandler<
                        RouteState,
                        RenderResult,
                        RouteHandler
                      > | null
                    ]
                  >)
                : loadRoute(routes, index + 1)
          )
        );
      };

      return loadRoute(transition.routes, 0).pipe(
        mergeMap(
          ([routeName, routeParams, handler]) =>
            (handler
              ? of([routeName, routeParams, handler])
              : this.createNotFoundHandler(transition).pipe(
                  map(v => [routeName, routeParams, v])
                )) as Observable<
              [
                string,
                RouteParams,
                StateHandler<RouteState, RenderResult, RouteHandler>
              ]
            >
        )
      );
    };

    const activateLoaded = ([route, params, handler]: [
      string,
      RouteParams,
      StateHandler<RouteState, RenderResult, RouteHandler>
    ]): Observable<RenderResult> => {
      this.activeRoute = [route, params, handler];
      // @ts-ignore
      return this.activeRoute[2].render();
    };

    this.resultsSubscription = merge(historyTransition$, navigateTransition$)
      .pipe(
        switchMap(transitionFromLocation),
        // transition handling
        map(matchRoutes),
        switchMap(loadMatched),
        switchMap(activateLoaded)
      )
      .subscribe(
        v => {
          this.renderResult$.next(v);
        },
        e => {
          if (this.renderResult$.observers.length) {
            this.renderResult$.error(e);
          } else {
            // tslint:disable-next-line no-console
            console.error(e);
          }
        }
      );
  }

  public stop(): void {
    if (this.resultsSubscription) {
      this.resultsSubscription.unsubscribe();
    }
  }

  public renderResult(): Observable<RenderResult> {
    return this.renderResult$.asObservable();
  }

  public isActive(route: string, params: RouteParams = {}): boolean {
    const [activeRoute, activeRouteParams] = this.activeRoute;
    if (!activeRoute || activeRoute.substr(0, route.length) !== route) {
      return false;
    }

    let paramName;

    const activeRouteInstance = this.routeCollection.getByName(activeRoute);
    const normalizedParams = normalizeParams(
      activeRouteInstance ? activeRouteInstance.searchParams : [],
      params
    );
    for (paramName in normalizedParams) {
      if (
        Object.prototype.hasOwnProperty.call(normalizedParams, paramName) &&
        normalizedParams[paramName] !== activeRouteParams[paramName]
      ) {
        return false;
      }
    }
    return true;
  }

  public createUrl(
    name: string,
    params: RouteParams = {},
    hash: string = ''
  ): string {
    const route = this.routeCollection.getByName(name);
    if (route) {
      const pathname = route.generatePath({
        ...this.activeRoute[1],
        ...params,
      });
      const search = generateQuery(params, route.searchParams);
      return this.history.createUrl(pathname, search, hash);
    }
    /* istanbul ignore else */
    if (process.env.NODE_ENV !== 'production') {
      // tslint:disable-next-line no-console
      console.error(`Route "${name}" not found`);
      return `#route-${name}-not-found`;
    }
    return '#';
  }

  public navigate(
    route: string,
    params: RouteParams = {},
    hash: string = '',
    state: object = {},
    source: NavigateSource = 'push'
  ): void {
    const url = this.createUrl(route, params, hash);
    this.navigateToUrl(url, state, source);
  }

  public navigateToUrl(
    url: string,
    state: object = {},
    source: NavigateSource = 'push'
  ): void {
    this.navigate$.next({ url, state, source });
  }
}
