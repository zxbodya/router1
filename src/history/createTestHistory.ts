import { Observable, Subject } from 'rxjs';
import { publishReplay, refCount, startWith } from 'rxjs/operators';

import { locationFromUrl } from '../utils/locationFromUrl';

import { History, Location } from './history';

type Navigate = (url: string, state?: object | null) => void;
type TestCallback = (
  action: 'push' | 'replace',
  historyState: {
    url: string;
    state?: object | null;
    title?: string | null;
  }
) => void;

export function createTestHistory(
  initialUrl: string,
  cb?: TestCallback
): History & { navigate: Navigate } {
  const location$: Subject<Location> = new Subject();
  return {
    createUrl(path: string, search?: string, hash?: string) {
      return `${path}${search ? `?${search}` : ''}${hash ? `#${hash}` : ''}`;
    },
    parseUrl(url: string) {
      return locationFromUrl(url);
    },
    push(
      url: string,
      state: object | null = null,
      title: string | null = null
    ) {
      if (cb) {
        cb('push', { url, state, title });
      }
    },
    replace(
      url: string,
      state: object | null = null,
      title: string | null = null
    ) {
      if (cb) {
        cb('replace', { url, state, title });
      }
    },
    navigate(url: string, state?: object | null) {
      location$.next({ ...locationFromUrl(url, state), source: 'pop' });
    },
    location: location$.pipe(
      startWith({ ...locationFromUrl(initialUrl), source: 'init' } as Location),
      publishReplay(1),
      refCount()
    ) as Observable<Location>,
  };
}
