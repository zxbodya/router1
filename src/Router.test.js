import { Router } from './Router';
import { Observable, helpers } from 'rx';
import { createTestHistory } from './createTestHistory';

const createTestHandler = (options = {}) =>
  (transition, route = { name: null, handlers: [] }, params = {}) =>
    ({
      load: () => Promise.resolve(true),
      hashChange: options.hashChange || helpers.noop,
      onBeforeUnload: options.onBeforeUnload || (() => ''),
      render() {
        const { location } = transition;
        return Observable.return({
          route: route.name,
          handlers: route.handlers,
          params,
          location,
        });
      },
    });

describe('Router', () => {
  it('works with empty route collection', (done) => {
    const history = createTestHistory('/');
    const router = new Router({
      routes: [],
      history,
      createNotFoundHandler: createTestHandler(),
      createHandler: createTestHandler(),
    });

    const t = process.env.NODE_ENV;
    process.env.NODE_ENV = 'development';
    expect(router.createUrl('main', {}, '')).toEqual('#route-main-not-found');
    process.env.NODE_ENV = 'production';
    expect(router.createUrl('main', {}, '')).toEqual('#');
    process.env.NODE_ENV = t;

    expect(router.isActive('main')).toEqual(false);

    router.renderResult().first().subscribe(renderResult => {
      expect(renderResult.route).toEqual(null);
      expect(renderResult.params).toEqual({});
      expect(renderResult.handlers).toEqual([]);
      expect(renderResult.location).toEqual({ pathname: '/', search: '', hash: '', state: {}, source: 'init' });

      expect(router.isActive('main')).toEqual(false);
      expect(router.activeRoute[0]).toEqual(null);
      expect(router.activeRoute[1]).toEqual({});
    }, () => {
    }, () => {
      done();
    });
  });

  it('matches route', (done) => {
    const history = createTestHistory('/');
    const router = new Router({
      routes: [{
        name: 'main',
        handler: 'main',
        url: '/',
      }],
      history,
      createNotFoundHandler: createTestHandler(),
      createHandler: createTestHandler(),
    });

    expect(router.createUrl('main', {}, '')).toEqual('/');
    expect(router.createUrl('main2', {}, '')).toEqual('#route-main2-not-found');

    router.renderResult().first().subscribe(renderResult => {
      expect(renderResult.route).toEqual('main');
      expect(renderResult.params).toEqual({});
      expect(renderResult.handlers).toEqual(['main']);
      expect(renderResult.location).toEqual({ pathname: '/', search: '', hash: '', state: {}, source: 'init' });
    }, () => {
    }, () => {
      expect(router.isActive('main')).toEqual(true);
      done();
    });
  });

  it('matches route with params', (done) => {
    const history = createTestHistory('/123');
    const router = new Router({
      routes: [{
        name: 'main',
        handler: 'main',
        url: '/<page:\\d+>',
      }],
      history,
      createNotFoundHandler: createTestHandler(),
      createHandler: createTestHandler(),
    });

    expect(router.createUrl('main', { page: 1000 }, '')).toEqual('/1000');

    router.renderResult().first().subscribe(renderResult => {
      expect(renderResult.route).toEqual('main');
      expect(renderResult.params).toEqual({ page: '123' });
      expect(renderResult.handlers).toEqual(['main']);
      expect(renderResult.location).toEqual({ pathname: '/123', search: '', hash: '', state: {}, source: 'init' });
    }, () => {
    }, () => {
      expect(router.isActive('main')).toEqual(true);
      expect(router.isActive('main', { page: '123' })).toEqual(true);
      expect(router.isActive('main', { page: '321' })).toEqual(false);
      done();
    });
  });

  it('matches route with params and search query and hash', (done) => {
    const history = createTestHistory('/123?q=text#anchor');
    const router = new Router({
      routes: [{
        name: 'main',
        handler: 'main',
        url: '/<page:\\d+>?q',
      }],
      history,
      createNotFoundHandler: createTestHandler(),
      createHandler: createTestHandler(),
    });

    expect(router.createUrl('main', { page: 1000, q: 1234 }, '')).toEqual('/1000?q=1234');
    expect(router.createUrl('main', { page: 1000, q: 1234 }, 'anchor')).toEqual('/1000?q=1234#anchor');

    router.renderResult().first().subscribe(renderResult => {
      expect(renderResult.route).toEqual('main');
      expect(renderResult.params).toEqual({ page: '123', q: 'text' });
      expect(renderResult.handlers).toEqual(['main']);
      expect(renderResult.location).toEqual({
        pathname: '/123',
        search: 'q=text',
        hash: 'anchor',
        state: {},
        source: 'init',
      });
    }, () => {
    }, () => {
      expect(router.isActive('main')).toEqual(true);
      expect(router.isActive('main', { page: '123', q: 'text' })).toEqual(true);
      expect(router.isActive('main', { page: '123', q: 'text' })).toEqual(true);
      expect(router.isActive('main', { page: '123', q: 'tex' })).toEqual(false);
      expect(router.isActive('main', { page: '12', q: 'text' })).toEqual(false);
      done();
    });
  });

  it('navigates', (done) => {
    const history = createTestHistory('/123?q=text#anchor');

    let hashChangeCount = 0;
    const router = new Router({
      routes: [
        {
          name: 'main',
          handler: 'main',
          url: '/<page:\\d+>?q',
        },
        {
          name: 'main1',
          handler: 'main1',
          url: '/m2/<page:\\d+>?q',
        },
      ],
      history,
      createNotFoundHandler: createTestHandler(),
      createHandler: createTestHandler({
        hashChange(location) {
          if (hashChangeCount === 0) {
            // router.navigate
            expect(location).toEqual({ pathname: '/123', search: 'q=text', hash: 'anc', state: {}, source: 'push' });
          }
          if (hashChangeCount === 1) {
            // onpopstate
            expect(location).toEqual({ pathname: '/123', search: 'q=text', hash: 'anc2', state: {}, source: 'pop' });
          }
          hashChangeCount += 1;
        },
      }),
    });

    let count = 0;

    router.renderResult().take(3).subscribe(renderResult => {
      if (count === 0) {
        expect(renderResult.route).toEqual('main');
        expect(renderResult.params).toEqual({ page: '123', q: 'text' });
        expect(renderResult.handlers).toEqual(['main']);
        expect(renderResult.location).toEqual({
          pathname: '/123',
          search: 'q=text',
          hash: 'anchor',
          state: {},
          source: 'init',
        });


        expect(router.isActive('main')).toEqual(true);
        expect(router.isActive('main', { page: '123', q: 'text' })).toEqual(true);
        expect(router.isActive('main', { page: '123', q: 'text' })).toEqual(true);
        expect(router.isActive('main', { page: '123', q: 'tex' })).toEqual(false);
        expect(router.isActive('main', { page: '12', q: 'text' })).toEqual(false);
      }

      if (count === 1) {
        expect(renderResult.route).toEqual('main1');
        expect(renderResult.params).toEqual({ page: '123', q: false });
        expect(renderResult.handlers).toEqual(['main1']);
        expect(renderResult.location).toEqual({ pathname: '/m2/123', search: '', hash: '', state: {}, source: 'push' });

        expect(router.isActive('main1')).toEqual(true);
        expect(router.isActive('main1', { page: '123' })).toEqual(true);
        expect(router.isActive('main1', { page: '123', q: false })).toEqual(true);
        expect(router.isActive('main1', { page: '123', q: 'tex' })).toEqual(false);
        expect(router.isActive('main1', { page: '12', q: 'text' })).toEqual(false);
      }
      if (count === 2) {
        expect(renderResult.route).toEqual('main');
        expect(renderResult.params).toEqual({ page: '123', q: 'text' });
        expect(renderResult.handlers).toEqual(['main']);
        expect(renderResult.location).toEqual({
          pathname: '/123',
          search: 'q=text',
          hash: 'anc',
          state: { a: true },
          source: 'push',
        });


        expect(router.isActive('main')).toEqual(true);
        expect(router.isActive('main', { page: '123', q: 'text' })).toEqual(true);
        expect(router.isActive('main', { page: '123', q: 'text' })).toEqual(true);
        expect(router.isActive('main', { page: '123', q: 'tex' })).toEqual(false);
        expect(router.isActive('main', { page: '12', q: 'text' })).toEqual(false);
      }
      count += 1;
    }, () => {
    }, () => {
      done();
    });
    setTimeout(() => {
      router.navigate('main', { page: '123', q: 'text' }, 'anc');
      history.navigate('/123?q=text#anc2', {});
      router.navigate('main1', { page: '123' });
      router.navigate('main', { page: '123', q: 'text' }, 'anc', { a: true });
    }, 10);
  });


  it('uses onBeforeUnload', (done) => {
    const history = createTestHistory('/');

    global.confirm = txt => txt === 'yes';

    let answer = 'no';

    const testHandlerOptions = {
      onBeforeUnload: () => answer,
    };
    const router = new Router({
      routes: [{
        name: 'main',
        handler: 'main',
        url: '/',
      }, {
        name: 'main2',
        handler: 'main2',
        url: '/2',
      }],
      history,
      createNotFoundHandler: createTestHandler(testHandlerOptions),
      createHandler: createTestHandler(testHandlerOptions),
    });

    router.start();

    setTimeout(() => {
      answer = 'no';
      router.navigate('main2', {});
    }, 5);

    setTimeout(() => {
      expect(router.isActive('main')).toEqual(true);
      answer = 'no';
      history.navigate('/2');
    }, 10);

    setTimeout(() => {
      expect(router.isActive('main')).toEqual(true);
      answer = 'yes';
      history.navigate('/2');
    }, 20);

    setTimeout(() => {
      expect(router.isActive('main2')).toEqual(true);
      answer = 'yes';
      router.navigate('main', {});
    }, 30);

    setTimeout(() => {
      expect(router.isActive('main')).toEqual(true);
      router.stop();
      done();
    }, 40);
  });

  it('handles transition.forward calls', (done) => {
    const history = createTestHistory('/');

    const createHandler = (transition, route = { name: null, handlers: [] }) =>
      ({
        load: () => Promise.resolve(true),
        hashChange: helpers.noop,
        onBeforeUnload: () => '',
        render() {
          return Observable.return(route.handlers[0](transition));
        },
      });

    const router = new Router({
      routes: [
        {
          name: 'main',
          handler: () => 'main',
          url: '/',
        },
        {
          name: 'main2',
          handler: t => t.forward('/3'),
          url: '/2',
        },
        {
          name: 'main3',
          handler: () => 'main3',
          url: '/3',
        },
        {
          name: 'main4',
          handler: t => (t.location.hash === '' ? t.forward('/4#123') : ''),
          url: '/4',
        },
      ],
      history,
      createNotFoundHandler: createHandler,
      createHandler,
    });

    router.start();

    setTimeout(() => {
      expect(router.isActive('main')).toEqual(true);
      router.navigate('main2', {});
    }, 5);

    setTimeout(() => {
      expect(router.isActive('main3')).toEqual(true);
      router.navigate('main4', {});
    }, 10);

    setTimeout(() => {
      expect(router.isActive('main4')).toEqual(true);
      expect(router.currentLocation.hash).toEqual('123');
      router.stop();
      done();
    }, 15);
  });

  it('crashes when transition.forward navigates to same page', (done) => {
    const history = createTestHistory('/');

    const createHandler = (transition, route = { name: null, handlers: [] }) =>
      ({
        load: () => Promise.resolve(true),
        hashChange: helpers.noop,
        onBeforeUnload: () => '',
        render() {
          return Observable.return(route.handlers[0](transition));
        },
      });

    const router = new Router({
      routes: [
        {
          name: 'main',
          handler: () => 'main',
          url: '/',
        },
        {
          name: 'main2',
          handler: t => t.forward('/2'),
          url: '/2',
        },
      ],
      history,
      createNotFoundHandler: createHandler,
      createHandler,
    });

    let hasError = false;
    router.renderResult().subscribeOnError(() => {
      hasError = true;
    });

    setTimeout(() => {
      expect(router.isActive('main')).toEqual(true);
      router.navigate('main2', {});
    }, 5);

    setTimeout(() => {
      expect(hasError).toEqual(true);
      router.stop();
      done();
    }, 10);
  });


  it('crashes when transition.forward calls cause redirect loop', (done) => {
    const history = createTestHistory('/');

    const createHandler = (transition, route = { name: null, handlers: [] }) =>
      ({
        load: () => Promise.resolve(true),
        hashChange: helpers.noop,
        onBeforeUnload: () => '',
        render() {
          return Observable.return(route.handlers[0](transition));
        },
      });

    const router = new Router({
      routes: [
        {
          name: 'main',
          handler: () => 'main',
          url: '/',
        },
        {
          name: 'main2',
          handler: t => t.forward('/3'),
          url: '/2',
        },
        {
          name: 'main3',
          handler: t => t.forward('/2'),
          url: '/3',
        },
      ],
      history,
      createNotFoundHandler: createHandler,
      createHandler,
    });

    let hasError = false;
    router.renderResult().subscribeOnError(() => {
      hasError = true;
    });

    setTimeout(() => {
      expect(router.isActive('main')).toEqual(true);
      router.navigate('main2', {});
    }, 5);

    setTimeout(() => {
      expect(hasError).toEqual(true);
      router.stop();
      done();
    }, 30);
  });
});
