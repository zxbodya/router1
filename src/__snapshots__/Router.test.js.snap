// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Router logs exception if not observer for it 1`] = `[Error: state rendering failed]`;
